File: output_repair/claude-opus/sample_5/repair_3/c/c/Struct/spouse.c
#include "stdlib.h"

struct person
{
  char *name;
  struct person *spouse;
};

/*@
predicate person(struct person *p;) =
  p->name |-> _ &*& p->spouse |-> _;
@*/

struct person *create_person()
//@ requires true;
//@ ensures person(result) &*& result->spouse |-> 0;
{
  struct person *p = malloc(sizeof(struct person));
  if (p == 0)
    abort();
  p->spouse = 0;
  //@ open person(p);
  //@ close person(p);
  return p;
}

void marry(struct person *this, struct person *other)
//@ requires person(this) &*& person(other);
//@ ensures person(this) &*& person(other) &*& this->spouse |-> other &*& other->spouse |-> this;
{
  //@ open person(this);
  //@ open person(other);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this);
  //@ close person(other);
}

struct person *person_get_spouse(struct person *this)
//@ requires person(this);
//@ ensures person(this) &*& this->spouse |-> result;
{
  //@ open person(this);
  struct person *result = this->spouse;
  //@ close person(this);
  return result;
}

void divorce(struct person *this, struct person *other)
//@ requires person(this) &*& person(other) &*& this->spouse |-> other &*& other->spouse |-> this;
//@ ensures person(this) &*& person(other) &*& this->spouse |-> 0 &*& other->spouse |-> 0;
{
  //@ open person(this);
  //@ open person(other);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close person(this);
  //@ close person(other);
}

void die(struct person *this)
//@ requires person(this) &*& this->spouse |-> 0;
//@ ensures true;
{
  //@ open person(this);
  if (this->spouse != 0)
  {
    this->spouse->spouse = 0;
  }
  free(this);
}

int main()
//@ requires true;
//@ ensures true;
{
  struct person *alice = create_person();
  struct person *bob = create_person();
  struct person *eve = 0;
  //@ open person(alice);
  //@ open person(bob);
  //@ close person(alice);
  //@ close person(bob);
  marry(alice, bob);
  eve = create_person();
  //@ open person(eve);
  //@ close person(eve);
  divorce(bob, alice);
  marry(bob, eve);
  //@ open person(alice);
  //@ close person(alice);
  die(alice);
  //@ open person(bob);
  //@ open person(eve);
  //@ close person(bob);
  //@ close person(eve);
  divorce(bob, eve);
  die(bob);
  die(eve);
  return 0;
}

File: output_20260321_064953/gpt52/java/Linklist/Spouse.java
class Person {

    protected Person spouse;

    /*@
    predicate PersonInv(Person p; Person s) =
        p.spouse |-> s;
    @*/

    
    public  void spouse_symm()
        
        
    //@ requires PersonInv(this, ?s) &*& s != null &*& PersonInv(s, this);
    //@ ensures PersonInv(this, s) &*& PersonInv(s, this);
    {
        //@ open PersonInv(this, s);
        //@ open PersonInv(s, this);
        //@ close PersonInv(s, this);
        //@ close PersonInv(this, s);
    }

    public Person()
        
        
    //@ requires true;
    //@ ensures PersonInv(this, null);
    {
        //@ close PersonInv(this, null);
    }
    
    public Person getSpouse()
        
        
    //@ requires PersonInv(this, ?s);
    //@ ensures PersonInv(this, s) &*& result == s;
    {
        //@ open PersonInv(this, s);
        Person result = spouse;
        //@ close PersonInv(this, s);
        return result;
        
    }
    
    protected void setSpouse(Person other)
        
        
    //@ requires PersonInv(this, ?s0) &*& PersonInv(other, ?s1);
    //@ ensures PersonInv(this, other) &*& PersonInv(other, this);
    {
        //@ open PersonInv(this, s0);
        //@ open PersonInv(other, s1);
        spouse = other;
        other.spouse = this;
        //@ close PersonInv(other, this);
        //@ close PersonInv(this, other);
        
    }
    
    protected void clearSpouse()
        
        
    //@ requires PersonInv(this, ?s) &*& s != null &*& PersonInv(s, this);
    //@ ensures PersonInv(this, null) &*& PersonInv(s, null);
    {
        //@ open PersonInv(this, s);
        //@ open PersonInv(s, this);
        spouse.spouse = null;
        spouse = null;
        //@ close PersonInv(s, null);
        //@ close PersonInv(this, null);
        
    }
    
    void marry(Person other)
        
        
    //@ requires PersonInv(this, ?s0) &*& PersonInv(other, ?s1);
    //@ ensures PersonInv(this, other) &*& PersonInv(other, this);
    {
        other.setSpouse(this);
        
    }
    
    void divorce()
        
        
    //@ requires PersonInv(this, ?s) &*& s != null &*& PersonInv(s, this);
    //@ ensures PersonInv(this, null) &*& PersonInv(s, null);
    {
        spouse.clearSpouse();
        
    }

}

class Program {

    public static void main(String[] args)
        
        
    //@ requires true;
    //@ ensures true;
    {
        Person a = new Person();
        Person b = new Person();
        //@ open Person.PersonInv(a, null);
        //@ open Person.PersonInv(b, null);
        a.marry(b);
        //@ open Person.PersonInv(a, b);
        //@ open Person.PersonInv(b, a);
        b.divorce();
        //@ close Person.PersonInv(a, null);
        //@ close Person.PersonInv(b, null);
    }

}